modules/substrate/src/verifier.rs

```diff
@@ -127,22 +134,39 @@ where
 			return Err(ImportError::InvalidChildNumber);
 		}
 
-		// This header requires a justification since it enacts an authority set change. We don't
-		// need to act on it right away (we'll update the set once this header gets finalized), but
+		// A header requires a justification if it enacts an authority set change. We don't
+		// need to act on it right away (we'll update the set once the header gets finalized), but
 		// we need to make a note of it.
 		//
-		// TODO: This assumes that we can only have one authority set change pending at a time.
-		// This is not strictly true as Grandpa may schedule multiple changes on a given chain
-		// if the "next next" change is scheduled after the "delay" period of the "next" change
-		let requires_justification = if let Some(change) = self.storage.scheduled_set_change() {
-			change.height == *header.number()
+		// Note: This assumes that we can only have one authority set change pending per fork at a
+		// time. While this is not strictly true of Grandpa (it can have multiple pending changes,
+		// even across forks), this assumption simplifies our tracking of authority set changes.
+		let mut signal_hash = parent_header.signal_hash;
+		let scheduled_change = find_scheduled_change(&header);
+
+		// Check if our fork is expecting an authority set change
+		let requires_justification = if let Some(hash) = signal_hash {
+			if let Some(pending_change) = self.storage.scheduled_set_change(hash) {
```

Comment:
nit: multiple nested `if`s are hard to read, especially when they're accompanied with `else`s. Since you have `unreachable!()` in `else` branch anyway, maybe just change to `.expect()` here?